icu4c/source/i18n/unitconverter.cpp

```diff
@@ -5,14 +5,495 @@
 
 #if !UCONFIG_NO_FORMATTING
 
+#include <cmath>
+
+#include "charstr.h"
+#include "double-conversion.h"
+#include "measunit_impl.h"
+#include "resource.h"
+#include "uassert.h"
+#include "unicode/stringpiece.h"
+#include "unicode/unistr.h"
+#include "unicode/utypes.h"
+#include "unitconverter.h"
+#include "uresimp.h"
+
 U_NAMESPACE_BEGIN
 
+namespace {
+
+using icu::double_conversion::StringToDoubleConverter;
+
+/* Internal Structure */
+
+// Represents a raw convert unit.
+struct ConvertUnit {
+    StringPiece source;
+    StringPiece target;
+
+    // NOTE
+    //  If the source and target are equal, the factor will be "1"
+    StringPiece factor = "1";
+
+    // NOTE
+    //  If the offset is not exist, its value will be "0"
+    StringPiece offset = "0";
+
+    // NOTE
+    //  If reciprocal is not exits, its value will be `false`
+    bool reciprocal = false;
+};
+
+typedef enum Signal {
+    NEGATIVE = -1,
+    POSITIVE = 1,
+} Signal;
+
+enum UnitsCase {
+    RECIPROCAL,
+    CONVERTIBLE,
+    UNCONVERTIBLE,
+};
+
+/* Helpers */
+
+// Returns `double` from a scientific number(i.e. "1", "2.01" or "3.09E+4")
+double strToDouble(StringPiece strNum) {
+    std::string charNum;
+    for (int i = 0; i < strNum.length(); i++) {
+        charNum += strNum.data()[i];
+    }
+
+    // We are processing well-formed input, so we don't need any special options to
+    // StringToDoubleConverter.
+    StringToDoubleConverter converter(0, 0, 0, "", "");
+    int32_t count;
+    return converter.StringToDouble(reinterpret_cast<const uint16_t *>(strNum.length()), strNum.length(),
+                                    &count);
+}
+
+// Returns `double` from a scientific number that could has a division sign (i.e. "1", "2.01", "3.09E+4"
+// or "2E+2/3")
+double strHasDivideSignToDouble(StringPiece strWithDivide, UErrorCode &status) {
+    CharString charNum(strWithDivide, status);
+    if (U_FAILURE(status)) return 0.0;
+
+    int divisionSignInd = -1;
+    for (int i = 0, n = charNum[i]; i < n; ++i) {
```

Comment:
This doesn't look right to me - the first parameter of a for loop is the initialiser, this is setting n to charNum[0] as though the first character of the strWithDivide is indicating its length? (Pascal-style strings, rather than C-style null-terminated strings, maybe?)

Reply:
For a char*, the for-loop could be this:

    for (int i = 0; charNum[i] != 0; ++i) {

Iterate until a null-termination character is found.

For StringPiece, it should probably be strWithDivide.length(). For CharString charNum, charNum.length(). But why create a copy of the StringPiece (charNum)? I think a for-loop to find a "/" in strWithDivide could look like this:

    for (int i = 0; i < strWithDivide.length(); ++i) {
        if (strWithDivide.data()[i] == '/') {
            divisionSignInd = i;
            break;
        }
    }